Expand requirements for Python (#2649)
chainguard-dev/edu: content/chainguard/libraries/python/overview.md

managers, you can use Chainguard Libraries for Python as a secure source of
truth for your development process.

## Technical Details

Most organizations consume Chainguard Libraries for Python through a repository

> manager such as Cloudsmith, JFrog Artifactory or Sonatype Nexus Repository. For

Author: Done

full details, refer to our [Global Configuration
documentation](/chainguard/libraries/python/global-configuration/). The rest of
this article provides details of the underlying implementation of Chainguard

> Libraries for Python and how to access individual libraries manually.

Author: kinda .. but not really.. we make a distinction .. directly is using just browsing and simple scripts.. direct access is without a repo manager .. which is covered elsewhere

The Chainguard Libraries for Python indexes use the PyPI repository format and
only include release artifacts of the libraries built by Chainguard from source.

The URLs for the repositories are:

```
https://libraries.cgr.dev/python/
https://libraries.cgr.dev/python-remediated/
```

The first index provides all Python libraries that Chainguard builds from
source, without remediated versions. The second index provides remediated
libraries with high and critical CVE fixes applied to older versions. The

> separate indexes provide the option to make remediated versions available for

Reviewer: ```suggestion
separate indexes provide you with the option to make remediated versions available for
```

> your development or opt out of using these versions completely and continue to

Reviewer: ```suggestion
your development or to opt out of using these versions completely and continue
```

> use upstream versions only.

Author: Done

> use non-remediated versions only.

Reviewer: ```suggestion
using non-remediated versions only.
```

For example, `aiohttp` has fixes for both  CVE-2024-23334 and CVE-2024-30251 in
the version `3.9.1+cgr.2`.

<a id="requirements"></a>

## Runtime and Build Requirements

The runtime requirements for Python artifacts available from Chainguard
Libraries for Python are identical to the requirements of the original upstream
project. For example, if a Python wheel retrieved from PyPI requires Python 3.10
or higher, the same Python 3.10 runtime requirement applies to the binary
artifact from Chainguard Libraries for Python.

> All packages are distributed as Python wheels, following the [Python binary

Reviewer: ```suggestion
Chainguard distributes all packages as Python wheels, following the [Python binary
```

trying to cut out passive voice where I can, not sure this is the right move here though

distribution
format](https://packaging.python.org/en/latest/specifications/binary-distribution-format/)
and are typically accompanied by a source distribution tarball. For example, for
the package `flask` version `3.1.2` the files are the
`flask-3.1.2-py3-none-any.whl` wheel and the `flask-3.1.2.tar.gz` source
distribution tarball. Note that the wheel filename includes the indication
`none-any` that signals the platform independence of the wheel.

Some Python libraries include Python extensions that depend on native binaries
supplied by the operating system or included in the distribution archive. The
following list displays all available files for the `pyyaml` package version

> `6.0.3`. Includes numerous wheel files for different build environments and the

Author: Done

pyyaml-6.0.3-cp310-cp310-manylinux_2_39_x86_64.whl
pyyaml-6.0.3-cp311-cp311-manylinux_2_28_aarch64.whl
pyyaml-6.0.3-cp311-cp311-manylinux_2_28_x86_64.whl
pyyaml-6.0.3-cp311-cp311-manylinux_2_39_aarch64.whl
pyyaml-6.0.3-cp311-cp311-manylinux_2_39_x86_64.whl
pyyaml-6.0.3-cp312-cp312-manylinux_2_28_aarch64.whl
pyyaml-6.0.3-cp312-cp312-manylinux_2_28_x86_64.whl
pyyaml-6.0.3-cp312-cp312-manylinux_2_39_aarch64.whl
pyyaml-6.0.3-cp312-cp312-manylinux_2_39_x86_64.whl
pyyaml-6.0.3-cp313-cp313-manylinux_2_28_aarch64.whl
pyyaml-6.0.3-cp313-cp313-manylinux_2_28_x86_64.whl
pyyaml-6.0.3-cp313-cp313-manylinux_2_39_aarch64.whl
pyyaml-6.0.3-cp313-cp313-manylinux_2_39_x86_64.whl
pyyaml-6.0.3-cp39-cp39-manylinux_2_28_aarch64.whl
pyyaml-6.0.3-cp39-cp39-manylinux_2_28_x86_64.whl
pyyaml-6.0.3-cp39-cp39-manylinux_2_39_aarch64.whl
pyyaml-6.0.3-cp39-cp39-manylinux_2_39_x86_64.whl
pyyaml-6.0.3.tar.gz
```

The different file names use the following qualifiers to allow downloading the
correct wheel file by the build and packaging tool:

* The values `cp*` indicate CPython compatibility.
* The first value `cp39`, `cp310`, and others indicates the compatibility with
  Python version `3.9`, `3.10`, and others.

> * The second value `cp39`, `cp310`, and others indicates the compatibility with

Author: Done

  the Python Application Binary Interface (ABI) version `3.9`, `3.10`, and
  others.
* The `manylinux` value indicates that the wheel is suitable for any modern
  Linux distribution, since not distribution specific requirements are in place.
* The `2_28` and `2_39` values indicate suitability for environments using
  `glibc 2.28` and higher or `glibc 2.39` and higher.
* The `aarch64` and `x86_64` values indicated the processor architecture
  requirement for ARM or x86.

As a result of the provided files for these libraries, the following
requirements apply:

> * All Linux distributions must use glibc 2.28 or higher, such as RHEL 8, Ubuntu

Reviewer: ```suggestion
* All Linux distributions must use `glibc 2.28` or higher, such as RHEL 8, Ubuntu
```

consistency w/ above

  20.04, or Amazon Linux 2023.
* On Windows and MacOS you must use a suitable Linux distribution with a

>   container solution, such as WSL2, apple/container or Docker Desktop.

Reviewer: Would you want to link to some of these tools? I never heard of apple/container (#linux4life) 

* https://learn.microsoft.com/en-us/windows/wsl/
* https://apple.github.io/container/documentation/
* https://docs.docker.com/desktop/

> * Processor architecture of the runtime environment must be `x86_64` or

Reviewer: ```suggestion
* The runtime environment's processor architecture must be `x86_64` or
```

Author: Done

  `aarch64`.

### Behavior on Windows and MacOS

As a result of the requirements detailed in the preceding section, the developer

> experience on Windows and MacOS platforms is impacted. Packages that include

Reviewer: ```suggestion
experience on Windows and MacOS platforms is impacted. Chainguard only provides packages including native binaries for Linux platforms.
```

native binaries are provided for Linux platforms only. These are not suitable
for use on Windows or MacOS.

When using Chainguard Libraries for Python on these platforms,

> retrieval for these libraries and therefore application builds fail. The

Reviewer: ```suggestion
retrieval for these libraries, and therefore application builds, fail. The
```

Author: Done

suggested deployment, detailed in [Technical Details](#technical-details) and

> [Global Configuration](/chainguard/libraries/python/global-configuration/), uses

Reviewer: ```suggestion
[Global Configuration](/chainguard/libraries/python/global-configuration/), involves
```

not sure that's the right word, but should avoid using "uses" twice

Author: Thats good.

a repository manager that uses PyPI as a fall back for such packages. With this

> configuration any build on Windows or MacOS continues to work and pulls the

Reviewer: ```suggestion
configuration, any build on Windows or MacOS continues to work and pulls the
```

> package from PyPI, since it is not available from Chainguard. Any build on Linux

Reviewer: ```suggestion
package from PyPI, since it is not available from Chainguard. Any build on Linux,
```

> however uses suitable package from the Chainguard. When using CVE remediation,

Reviewer: ```suggestion
however, uses suitable a package from the Chainguard Library. When using CVE remediation,
```

i think?
